### PR TITLE
added uiEntry + fixed a wrong function call

### LINKED ISSUE #23: Remove unwanted casting in wrappers
In wrappers (i.e. the files under `pylibui/libui`) there are some bool casts. We want wrappers to be really simple. So these casts should be moved to corresponding files in `pylibui/controls` and docstrings must be corrected as well. 

* checkbox.py
* control.py
* window.py

Correct files above and please review all wrappers. If they have unnecessary details that can be implemented in `pylibui/controls`, fix them as well.

Feel free to ask for help if you need :octocat:

### PR DESCRIPTION
**Pull request by superzazu**
I added uiEntries + fixed a wrong function call in `pylibui/libui/checkbox.py`

In the file `examples/entry.py`, I temporarily commented `app.close()` to avoid crashes (and added a comment that links to the issue #18)

Note: the deletion of "raise" in tests is already done in another branch ([here](https://github.com/superzazu/pylibui/tree/remove_raises_in_tests)). I'll make more PR (this one among two others which fixes #23 & #24) when this one is merged

What do you guys think ?

### CODE REVIEW

File: pylibui/controls/entry.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+from .control import Control
+
+
+class BaseEntry(Control):
+
+    def setText(self, text):
+        """
+        Sets the text of the entry.
+
+        :param text: the text of the entry
```
Comment (nsgonultas) [inline pylibui/controls/entry.py]: What about making this just `string` instead of `the text of the entry`?

File: pylibui/controls/entry.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        Sets whether the entry is read only or not.
+
+        :param read_only: bool
+        :return: None
+        """
+        libui.uiEntrySetReadOnly(self.control, read_only)
+
+    def getReadOnly(self):
+        """
+        Sets whether the entry is read only or not.
```
Comment (nsgonultas) [inline pylibui/controls/entry.py]: This should be *Gets...*

File: pylibui/libui/entry.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+# - int uiEntryReadOnly(uiEntry *e);
+def uiEntryReadOnly(entry):
+    """
+    Returns whether the entry is read only or not.
+
+    :param entry: uiEntry
+    :return: string
+    """
+
+    return bool(clibui.uiEntryReadOnly(entry))
```
Comment (nsgonultas) [inline pylibui/libui/entry.py]: This cast should be in wrapper I think, and return value should be int instead of string, in docstring.

File: pylibui/libui/entry.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+
+# - void uiEntrySetReadOnly(uiEntry *e, int readonly);
+def uiEntrySetReadOnly(entry, read_only):
+    """
+    Sets whether the entry is read only or not.
+
+    :param entry: uiEntry
+    :param read_only: bool
```
Comment (nsgonultas) [inline pylibui/libui/entry.py]: `int` instead of `bool`. https://github.com/joaoventura/pylibui/pull/21/files#diff-28abfffa51d42558eda4a49aecd41b5fL60

File: pylibui/controls/entry.py (hunk 2 above)
Comment (nsgonultas) [inline pylibui/controls/entry.py]: Or *Returns* :grin:

File: tests/test_entry.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+from tests.utils import WindowTestCase
+
+
+class EntryTest(WindowTestCase):
+
+    def setUp(self):
+        super().setUp()
+        self.entry = Entry()
+        self.password_entry = PasswordEntry()
+        self.search_entry = SearchEntry()
```
Comment (nsgonultas) [inline tests/test_entry.py]: Did you used `password_entry` and `search_entry` in this file, am I overlooking them? If you didn't use I think we can remove them because they all share the same functions, am I wrong?

File: tests/test_entry.py [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+        self.entry.setReadOnly(True)
+        self.assertEqual(self.entry.getReadOnly(), True)
+
+    def test_read_only_can_be_set_to_false(self):
+        """Tests the entry's `read_only` attribute can be set to False."""
+        self.entry.setReadOnly(False)
+        self.assertEqual(self.entry.getReadOnly(), False)
+
+    def test_onChanged_is_called(self):
+        raise  # TODO: how do we simulate mouse press ?
```
Comment (nsgonultas) [inline tests/test_entry.py]: If we will remove all lines like this, why did you add one here, maybe a mistake?

File: pylibui/controls/entry.py (hunk 1 above)
Comment (superzazu) [inline pylibui/controls/entry.py]: Yes, you are right

File: pylibui/controls/entry.py (hunk 2 above)
Comment (superzazu) [inline pylibui/controls/entry.py]: Whoops

File: pylibui/libui/entry.py (hunk 3 above)
Comment (superzazu) [inline pylibui/libui/entry.py]: Yes you are right. The reason I did not do it is simply because I wrote the code before the issue was submitted. (24 Jun)
I can fix it on this PR; however I already did another branch -which I will submit in a PR after this one- [which removes all the casts in `pylibui/libui/*.py` files](https://github.com/superzazu/pylibui/commit/bb1dcfe7510c4f57ede1ca10ecc4c8f7aa2b9953). It's which one you prefer :-)

Concerning the "int" instead of "string", you are right

Comment (nsgonultas) [inline pylibui/libui/entry.py]: I see. It's okay, we can merge this one, then send the others :+1:

File: pylibui/libui/entry.py (hunk 4 above)
Comment (superzazu) [inline pylibui/libui/entry.py]: You are right

File: tests/test_entry.py (hunk 5 above)
Comment (superzazu) [inline tests/test_entry.py]: True, I do not need these two variables. I just instanced them here at first, but don't need it.

**Comment by superzazu** [comment]
Okay, thanks for the feedback (and sorry for the typos...), I'm going to do another commit :-)

As I said, concerning the two comments about" removing the raise in tests" + "putting the casts in controls"; it is because this code is a bit old.
I thought it'd be cleaner to remove them all at once in specific commits (already wrote these commits, will submit two PRs as soon as this one is merged). However, I can remove them in this PR if you'd like. :)

**Comment by nsgonultas** [comment]
No, it will be better if they come in another PR. Sorry for comments, I haven't fully understood why you're mentioning about other two PRs in this one, now it's clear :+1:

**Comment by nsgonultas** [comment]
BTW, what is your output after executing all test suite? It seems you've fixed the segmentation thing in checkbox, for me.

**Comment by superzazu** [comment]
On my mac, I never had any problems with segfaults while running tests (only when running examples). Here's the complete output:

```
.E.....E.........E.........E..E.........
======================================================================
ERROR: test_onClick_is_called_when_user_presses_button (tests.test_button.ButtonTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/nikkos/Documents/pylibui/tests/test_button.py", line 28, in test_onClick_is_called_when_user_presses_button
    raise  # TODO: how do we simulate mouse press ?
RuntimeError: No active exception to reraise

======================================================================
ERROR: test_onToggle_is_called_when_user_toggles_checkbox (tests.test_checkbox.CheckboxTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/nikkos/Documents/pylibui/tests/test_checkbox.py", line 42, in test_onToggle_is_called_when_user_toggles_checkbox
    raise  # TODO: how do we simulate mouse press ?
RuntimeError: No active exception to reraise

======================================================================
ERROR: test_onChanged_is_called (tests.test_entry.EntryTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/nikkos/Documents/pylibui/tests/test_entry.py", line 41, in test_onChanged_is_called
    raise  # TODO: how do we simulate mouse press ?
RuntimeError: No active exception to reraise

======================================================================
ERROR: test_onChanged_is_called_when_user_changes_sliders_value (tests.test_slider.SliderTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/nikkos/Documents/pylibui/tests/test_slider.py", line 29, in test_onChanged_is_called_when_user_changes_sliders_value
    raise  # TODO: how do we simulate mouse press ?
RuntimeError: No active exception to reraise

======================================================================
ERROR: test_onChanged_is_called_when_user_changes_spinbox_value (tests.test_spinbox.SpinboxTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/nikkos/Documents/pylibui/tests/test_spinbox.py", line 29, in test_onChanged_is_called_when_user_changes_spinbox_value
    raise  # TODO: how do we simulate mouse press ?
RuntimeError: No active exception to reraise

----------------------------------------------------------------------
Ran 40 tests in 0.170s

FAILED (errors=5)
```

**Comment by superzazu** [comment]
Erm, I'm dumb. :D Of course I did something to the checkboxes : https://github.com/joaoventura/pylibui/pull/27/commits/f60452d32ef059315b08b89bb1aad3c0e4450313

I was calling a wrong method (uiButtonOnClicked instead of uiCheckboxOnToggled). That may be the problem you were having while running tests.

I commited the fixes to these typos.

**Comment by nsgonultas** [comment]
Weird. This one is just testing control, FYI. And I still have an segfault about while testing window.

```
.....F
======================================================================
FAIL: test_visible_initial_value (tests.test_control.ControlTest)
Tests the control's `visible` initial value is True.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/enes/Code/GitHub/spzzpyl/tests/test_control.py", line 19, in test_visible_initial_value
    self.assertEqual(self.control.getVisible(), True)
AssertionError: 0 != True

----------------------------------------------------------------------
Ran 6 tests in 0.046s

FAILED (failures=1
```

**Comment by nsgonultas** [comment]
I saw your change and I'm now suspecting about another wrong call in window.

**Comment by superzazu** [comment]
Weird. I don't have this problem here: 
```
nikkos@debbie:pylibui $ p -m unittest tests/test_control.py              master
......
----------------------------------------------------------------------
Ran 6 tests in 0.068s

OK
```

I checked window files, I did not see anything obvious like the "wrong call" I just fixed.

**Comment by nsgonultas** [comment]
Let's focus on this PR and merge it, I will try to find the problem soon.

File: examples/entry.py [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+window = MyWindow('Entry example')
+window.setMargined(10)
+
+entry = MyEntry()
+search_entry = SearchEntry()
+password_entry = PasswordEntry()
+
+vbox = VerticalBox()
+vbox.setPadded(10)
```
Comment (nsgonultas) [inline examples/entry.py]: Why are you using 10 here? I thought it expects a bool value here, am I wrong?

Comment (superzazu) [inline examples/entry.py]: Yes, is it a mistake. I have to correct this

By the way, I think we should replace the occurrences of ".setMargined(1)" by ".setMargined(True)", and so on (as always, only in the pythonic wrapper). I'll open an issue for that.

**Comment by nsgonultas** [comment]
Thanks for PR :octocat:

**Comment by nsgonultas** [comment]
See, you have another 10 on the line window.setMargined, please fix it as
well.

22 Eki 2016 19:07 tarihinde "Nicolas Allemand" <notifications@github.com>
yazdı:

**Comment by superzazu** [comment]
The `setPadded` and `setMargined` calls have been corrected in #31 . I searched in the project (master) and did not find any occurrences to `setPadded(10)` and `setMargined(10)`. Maybe I missed something ?

**Comment by nsgonultas** [comment]
Sorry for that. I've sent that comment four days ago via Gmail. It took a bit late to appear here :expressionless: I was talking about line 30 in entry example.
